spec/components/plugin/instance_spec.rb

```diff
@@ -600,4 +600,55 @@ def register_locale(locale, opts)
       expect(ApiKeyScope.scope_mappings.dig(:groups, :create, :actions)).to contain_exactly(*actions)
     end
   end
+
+  describe '#add_directory_column' do
+    let!(:plugin) { Plugin::Instance.new }
+
+    before do
+      DirectoryItem.clear_plugin_queries
+    end
+
+    after do
+      DirectoryColumn.clear_plugin_directory_columns
+    end
+
+    describe "with valid column name" do
+      let(:column_name) { "random_c" }
+
+      before do
+        DB.exec("ALTER TABLE directory_items ADD COLUMN IF NOT EXISTS #{column_name} integer")
+      end
+
+      after do
+        DB.exec("ALTER TABLE directory_items DROP COLUMN IF EXISTS #{column_name}")
+      end
+
+      it 'creates a directory column record when directory items are refreshed' do
+        plugin.add_directory_column(column_name, query: "SELECT COUNT(*) FROM users", icon: 'recycle')
+        expect(DirectoryColumn.find_by(name: column_name, icon: 'recycle', enabled: false).present?).to be(false)
+
+        DirectoryItem.refresh!
+        expect(DirectoryColumn.find_by(name: column_name, icon: 'recycle', enabled: false).present?).to be(true)
+
+        # add_directory_column runs `DiscourseEvent.on("before_directory_refresh")`. The rails spec helper
+        # ensures that no test will run and change the total number of registrations. The easy solution is to
+        # manually clean up the registration:
+        DiscourseEvent.events.delete("before_directory_refresh")
```

Comment:
This feels a bit messy to me. Can you explain more about what you mean by "ensures that no test will run and change the total number of registrations". I would really prefer if we didn't have to delete stuff for tests to pass.

Reply:
This is the error with the line removed
```
1) Plugin::Instance#add_directory_column with valid column name creates a directory column record when directory items are refreshed
     Failure/Error: expect(before_event_count).to eq(after_event_count), "DiscourseEvent registrations were not cleaned up"
       DiscourseEvent registrations were not cleaned up
     # ./spec/rails_helper.rb:281:in `block (2 levels) in <top (required)>'
  
 ```
 
 Caused by https://github.com/discourse/discourse/blob/83a6ad32ffe75ae222028feddeca169fc5be54ac/spec/rails_helper.rb#L277-L282

Reply:
I see, thanks. I would prefer if you weren't reaching into `events` to delete manually and we had a method for this.

Normally a test calls `.off` to remove an event it wires up. Perhaps you could add `.all_off` and call that?

Reply:
Yeah, that's a really good solution. Pushed a commit that adds `DiscourseEvent#all_off`